File: solutions/cpp/Linked List Cycle II.cpp
#include <unordered_set>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        unordered_set<ListNode *> s;
        ListNode *temp = head;
        while (temp)
        {
            if (s.find(temp) != s.end())
                return temp;
            else
                s.insert(temp);
            temp = temp->next;
        }
        return NULL;
    }
};

File: solutions/python/Linked List Cycle II.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        nodes=set()
        temp=head;
        while temp:
            if temp not in nodes:
                nodes.add(temp)
            else:
                return temp
            temp=temp.next;
        return None